send/o365_mu_process.pl

```diff
@@ -202,352 +152,331 @@
+	print "-------------------------------------\n";
+	print Dumper($newUsersCache);	
+	print "--------------------------------------\n";
+	print "ALL DATA ABOUT CACHE GROUPS TO PROCESS\n";
+	print "--------------------------------------\n";
+	print Dumper($newGroupsCache);
 }
 
-foreach my $groupRecord (@allGroups) {
-	print FILE_GROUPS_CACHE $groupRecord . "\n";
-}
-
-#copy new cache files to place with old cache files
-close FILE_USERS_CACHE or die "Could not close file $newUsersCache: $!\n";
-close FILE_GROUPS_CACHE or die "Could not close file $newGroupsCache: $!\n";
-copy( $newUsersCache, $lastStateOfUsersFilename );
-copy( $newGroupsCache, $lastStateOfGroupsFilename );
-
-exit $returnCode;
-
-#---------------------------------SUBS------------------------------------
-
-#Sub to start all threads
-sub startThreads {
-	for(1 .. $THREAD_COUNT) {
-		threads->create(\&processTasks);
+#-------------------------------------------------------------------------
+#-------------------------------MAIN CODE---------------------------------
+#-------------------------------------------------------------------------
+
+my $countOfUsersToProcess = keys %$usersToProcessStruc;
+#check all processed users and add them to the cache if they were updated ok
+if($countOfUsersToProcess > 0 && $DRY_RUN == 0) {
+	#we need to do this in chunks to prevent big queues on the server site
+	my @chunksOfUsersStruc = createChunksForStruc( $usersToProcessStruc, $MAX_USERS_CHUNK_SIZE ); 
+	my $chunkCounter = 0;
+	foreach my $chunk (@chunksOfUsersStruc) {
+		$chunkCounter++;
+		if($DEBUG>0) { print "INFO: Users chunk number - $chunkCounter\n"; }
+		#users to be updated as content for call
+		my $usersContent = getUsersContent($chunk);
+		my $usersResult = callServerForUpdate($usersContent) ;
+		my $usersResultMap = transformResultToMap($usersResult);
+		#check all processed users and add them to the cache if they were updated ok
+		foreach my $key (keys %$chunk) {
+			my $processedUser = $chunk->{$key};
+			my $resultUser = $usersResultMap->{$key};
+
+			#user was processed
+			if($resultUser) {
+				#check if there was any exception
+				my $exception = $resultUser->{'ExceptionMessage'};
+				#exception means a problem, return an error, do not add to the cache
+				if($exception) {
+					my $exceptionType = $resultUser->{'ExceptionType'};
+					print "WARNING - User with ID $key was not processed properly due to: " . $exceptionType . " -- " . $exception . "\n";
+					$failedUsers++;
+					$returnCode = 1;
+				} else {
+					#no exception means ok, add to the cache
+					addObjectToCache($newUsersCache, $key, $processedUser);
+					$successUsers++;
+					if($DEBUG>0) { print "+++ USER: $key\n"; }
+				}
+			} else {
+				#if user was not processed at all, it is also an error, do not add to the cache
+				print "WARNING - User with ID $key was timeouted (not processed).\n";
+				$failedUsers++;
+				$returnCode = 1;
+			}
+		}
+		#save only if this is hard run (not the DRY run)
```

Comment:
this will never be a dry run due to the condition `if($countOfUsersToProcess > 0 && $DRY_RUN == 0)`